File: LeetCode/Easy/Question_160/Question_160.c
#include <stdio.h>

// Definition for singly-linked list.
struct ListNode
{
    int val;
    struct ListNode *next;
};

struct ListNode *getIntersectionNode(struct ListNode *headA, struct ListNode *headB) {
    struct ListNode *cpyHeadA = headA;
    struct ListNode *cpyHeadB = headB;

    int aLen = 0;
    int bLen = 0;

    while( cpyHeadA != NULL )
    {
        aLen++;
        cpyHeadA = cpyHeadA->next;
    }

    while( cpyHeadB != NULL )
    {
        bLen++;
        cpyHeadB = cpyHeadB->next;
    }

    int lenDiff = aLen - bLen;
    cpyHeadA = headA;
    cpyHeadB = headB;

    while( lenDiff > 0 )
    {
        cpyHeadA = cpyHeadA->next;
        lenDiff--;
    }

    while( lenDiff < 0 )
    {
        cpyHeadB = cpyHeadB->next;
        lenDiff++;
    }

    while( cpyHeadA )
    {
        if( cpyHeadA == cpyHeadB )
        {
            return cpyHeadA;
        }

        cpyHeadA = cpyHeadA->next;
        cpyHeadB = cpyHeadB->next;
    }
    
    return NULL;
}

//  Beats 78.15% Runtime, 38ms
//  Beats 57.12% Memory, 13.9mb

File: LeetCode/Easy/Question_160/Question_160.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> ListNode:
        
        nodeSet = set()
        
        while headA is not None:
            nodeSet.add(headA)
            headA = headA.next
        
        while headB is not None:
            if headB in nodeSet:
                return headB
            headB = headB.next
        
        return None

# Beats 93.26% Runtime, 150ms
# Beats 16.44% Memory, 30.1mb